src/huggingface_hub/repository.py

```diff
@@ -160,32 +164,42 @@ def clone_from(self, repo_url: str, use_auth_token: Union[bool, str, None] = Non
         is_hf_id = len(repo_url.split("/")) <= 2
         api = HfApi()
 
+        if not is_hf_id and repo_url.split("/")[-3] == "datasets":
+            self.repo_type = "dataset"
+
         if token is not None:
+            user, valid_organisations = api.whoami(token)
             if is_hf_url:
-                organization, repo_id = repo_url.split("/")[-2:]
+                namespace, repo_id = repo_url.split("/")[-2:]
             elif is_hf_id:
                 if len(repo_url.split("/")) == 2:
                     # Passed <user>/<model_id> or <org>/<model_id>
-                    organization, repo_id = repo_url.split("/")[-2:]
+                    namespace, repo_id = repo_url.split("/")[-2:]
                 else:
                     # Passed <model_id>
-                    organization, repo_id = api.whoami(token)[0], repo_url
+                    namespace, repo_id = user, repo_url
+
+                repo_url = ENDPOINT
+
+                if self.repo_type == "dataset":
+                    repo_url += "/datasets"
```

Comment:
Might be worth doing the same for spaces as well !

Reply:
pattern elsewhere is something like
```
        if self.repo_type in REPO_TYPES_URL_PREFIXES:
            repo_id = REPO_TYPES_URL_PREFIXES[self.repo_type] + repo_id
```

Reply:
I'll re-use that pattern! Thanks

Reply:
Oh this is better indeed, thanks for the tip!